### PR TITLE
Add media breakpoint tracking

### PR DESCRIPTION
**Pull request by jgrund**
Add a new module that will track current breakpoint size via the window resize event.

This can be used to handle responsive rendering changes in Rust code.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1519)

### CODE REVIEW

**Comment by jgrund** [comment]
This also fixes a bug where after clicking the hamburger button with the md breakpoint, the menu would not show with the lg breakpoint.

resize

File: iml-gui/crate/src/breakpoints.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        .as_f64()
+        .expect("Could not parse inner_width to f64")
+}
+
+/// Returns *maximum* matching breakpoint based on `window.inner_width`.
+/// This models the behavior of breakpoints in tailwind: https://tailwindcss.com/docs/responsive-design/
+pub(crate) fn size() -> Size {
+    let w = inner_width();
+
+    if w >= XL {
```
Comment (johnsonw) [inline iml-gui/crate/src/breakpoints.rs]: Any reason not to use a pattern match?

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: How? It's doing an ordering comparision, not an equal one.

Comment (johnsonw) [reply iml-gui/crate/src/breakpoints.rs]: meh, you could use a match guard but I guess it doesn't buy anything.

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: Yeah, I thought that might be what you meant. That would look noisier than just an if / else imo.

Comment (johnsonw) [reply iml-gui/crate/src/breakpoints.rs]: yes, I agree.

File: iml-gui/crate/src/breakpoints.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+use seed::window;
+
+pub(crate) const SM: f64 = 569.;
+pub(crate) const MD: f64 = 769.;
+pub(crate) const LG: f64 = 1025.;
+pub(crate) const XL: f64 = 1701.;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) enum Size {
+    XS,
```
Comment (ip1981) [inline iml-gui/crate/src/breakpoints.rs]: I wonder if this could be `enum Size { XS = 0, SM = 569,...}` and so on.

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: Hrrm, didn't consider that, looking.

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: So, this *does* work:

```rust
#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
pub(crate) enum Size {
    XS = 0,
    SM = 569,
    MD = 769,
    LG = 1025,
    XL = 1701,
}
```

However any usage of the enum then needs to be double casted to compare with the f64 of `inner_width`:

`w >= Size::XL as u64 as f64`

Not sure it's worth the cost of callers having to do that each time over the current `const` bindings.

Comment (ip1981) [reply iml-gui/crate/src/breakpoints.rs]: Well, then update `inner_width` to return an int. Width is integer after all.

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: yeah, ok.

File: iml-gui/crate/src/breakpoints.rs [hunk 3]
```diff
@@ -6,40 +6,35 @@
...
+        .expect("Could not parse inner_width to f64") as u64
 }
 
 /// Returns *maximum* matching breakpoint based on `window.inner_width`.
 /// This models the behavior of breakpoints in tailwind: https://tailwindcss.com/docs/responsive-design/
 pub(crate) fn size() -> Size {
     let w = inner_width();
 
-    if w >= XL {
+    if w >= Size::XL as u64 {
```
Comment (ip1981) [inline iml-gui/crate/src/breakpoints.rs]: Doesn't it work without type casting? O_o

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: Nope

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: Which is what I was referring to earlier. However, the compiler / rust-analyzer will tell you if you don't cast, so I think it's ok.

Comment (ip1981) [reply iml-gui/crate/src/breakpoints.rs]: I feel something like regret :)

Comment (ip1981) [reply iml-gui/crate/src/breakpoints.rs]: Oh, maybe use `#[repr]` for the enum?

Comment (jgrund) [reply iml-gui/crate/src/breakpoints.rs]: Nope, still same issue with `repr` attribute.

**Pull request merged**